signature/jws/verifier.go

```diff
@@ -0,0 +1,302 @@
+	}
+}
+
+// Verify verifies the signature and returns the verified descriptor and
+// metadata of the signed artifact.
+func (v *Verifier) Verify(ctx context.Context, signature []byte, opts notation.VerifyOptions) (notation.Descriptor, notation.Metadata, error) {
+	// unpack envelope
+	sig, err := openEnvelope(signature)
+	if err != nil {
+		return notation.Descriptor{}, notation.Metadata{}, err
+	}
+
+	// verify signing identity
+	key, err := v.verifySigner(&sig.Signature)
+	if err != nil {
+		return notation.Descriptor{}, notation.Metadata{}, err
+	}
+
+	// verify JWT
+	claim, err := v.verifyJWT(sig.SerializeCompact(), key)
+	if err != nil {
+		return notation.Descriptor{}, notation.Metadata{}, err
+	}
+
+	// extract metadata
+	var identity string
+	if value := claim.SignedAttributes.Reserved["identity"]; value != nil {
+		var ok bool
+		if identity, ok = value.(string); !ok {
+			return notation.Descriptor{}, notation.Metadata{}, errors.New("attribute: invalid identity")
+		}
+	}
+
+	return claim.SubjectManifest, notation.Metadata{
+		Identity:   identity,
+		Attributes: claim.SignedAttributes.Custom,
+	}, nil
+}
+
+// verifySigner verifies the signing identity and returns the verification key.
+func (v *Verifier) verifySigner(sig *jwsutil.Signature) (*VerificationKey, error) {
+	var header unprotectedHeader
+	if err := json.Unmarshal(sig.Unprotected, &header); err != nil {
+		return nil, err
+	}
+
+	if header.KeyID != "" {
+		if len(header.CertChain) != 0 {
+			return nil, errors.New("ambiguous signing identities")
+		}
+		if key, ok := v.keys[header.KeyID]; ok {
+			return key, nil
+		}
+		return nil, fmt.Errorf("unknown key id: %s", header.KeyID)
+	} else if len(header.CertChain) != 0 {
+		return v.verifySignerFromCertChain(header.CertChain, header.TimeStampToken, sig.Signature)
+	}
+	return nil, errors.New("signing identity not found")
+}
+
+// verifySignerFromCertChain verifies the signing identity from the provided certificate
+// chain and returns the verification key. The first certificate of the certificate chain
+// contains the key, which used to sign the artifact.
+// Reference: RFC 7515 4.1.6 "x5c" (X.509 Certificate Chain) Header Parameter.
+func (v *Verifier) verifySignerFromCertChain(certChain [][]byte, timeStampToken []byte, encodedSig string) (*VerificationKey, error) {
+	// prepare for certificate verification
+	certs := make([]*x509.Certificate, 0, len(certChain))
+	for _, certBytes := range certChain {
+		cert, err := x509.ParseCertificate(certBytes)
+		if err != nil {
+			return nil, err
+		}
+		certs = append(certs, cert)
+	}
+	intermediates := x509.NewCertPool()
+	for _, cert := range certs[1:] {
+		intermediates.AddCert(cert)
+	}
+	verifyOpts := v.VerifyOptions
+	verifyOpts.Intermediates = intermediates
+	if len(verifyOpts.KeyUsages) == 0 {
+		verifyOpts.KeyUsages = []x509.ExtKeyUsage{x509.ExtKeyUsageCodeSigning}
+	}
+
+	// verify the signing certificate
+	checkTimestamp := v.EnforceExpiryValidation
```

Comment:
It looks the option is to check the signature was signed when the signing cert was still valid. Maybe the name can be updated to reflect the intention. Expiry makes me think it is to check if the cert has expired.

Reply:
This name is defined in https://github.com/notaryproject/notaryproject/discussions/98. We may need to change the name in the spec first.